Require zip code when entering school info on Regional Partner Contact Form

islemaster opened the pull request:
Requiring a zip code allows us to match contacts to regional partners so often, so that we don't have to do as much manual matching of messages to partners from the central office.

- School Type? Public
  - District found in dropdown
    - School omitted entirely (user skips over the school dropdown and the "other school" checkbox)
      **Zip code is not displayed or required.**
    - School found in dropdown
      **Zip code is not displayed or required**
    - Other school not listed above
      We display two fields: School name, School zip code
      **Zip code is required if a school name is entered**
  - Other district not listed above
    We display three fields: District name, School name, School zip code
    **Zip code is required if a school name is entered**
- School Type? Private
  We display two fields: School name, School zip code
  **Zip code is required**

### Prior work:
- [Regional Partner Contact Form updates](https://github.com/code-dot-org/code-dot-org/pull/25165)
- [Hotfix: Send Regional Partner Emails from allowed address](https://github.com/code-dot-org/code-dot-org/pull/25233)
- [Send RegionalPartnerContact email from partner@ (and tests!)](https://github.com/code-dot-org/code-dot-org/pull/25238)

poorvasingal commented:
I thought @tanyaparker was suggesting we require a zip code no matter what (unless user picked a school from the dropdown, since then we can figure out what the zip code is ourselves).

islemaster commented:
Changed to _always_ require zipcode.  Doing manual testing now.  PTAL!

islemaster commented:
Updated with new requirements again.  PTAL.

This time I've removed the validation unit tests, which were getting hard to follow, in favor of an expanded suite of controller integration tests using newly-generated test data I got by actually submitting the form and copying the payload out of my developer tools.  I think this is a better representation of the actual behavior of the form and gives me confidence that we're never requiring zip code in a scenario where's it's not possible for the user to provide it.

mehalshah commented on dashboard/test/controllers/api/v1/pd/regional_partner_contacts_controller_test.rb:
```diff
@@ -1,44 +1,206 @@
+    assert_valid_form FOUND_DISTRICT_OTHER_SCHOOL
+  end
+
+  test 'zip code is required for found district and other school' do
```
I don't think the controller is the right place for all this testing. It really belongs in the model since you are testing the validator. You also don't need to have one separate case for each field that you are testing. You can try to create a regional partner contact without any info and assert that it is not valid, and that all the error messages correspond to all the missing fields.

islemaster replied:
PTAL.

I moved my validation cases back into the model unit tests, left only happy-path cases in the controller integration test, and deduplicated my test data into factory traits.

As discussed offline, I'd prefer to keep one test case per field in these tests for a couple of reasons:

1. There's some weird dependent "required" behavior in this form.
2. This form only generates one 'generic' validation error for any missing school info.
